source/llvm/ASTNodeCodeGen.cpp

```diff
@@ -637,6 +639,159 @@ llvm::Value* ASTNodeCodeGen::delayExprCodeGen(const libsbml::ASTNode* ast)
+            ASTNode* times = new ASTNode(AST_TIMES);
+            ASTNode* stoich = new ASTNode(AST_NAME);
+            stoich->setName((reaction->getId() + ":" + name).c_str());
+            times->addChild(stoich);
+            times->addChild(reaction->getKineticLaw()->getMath()->deepCopy());
+            amtRateRef->addChild(times);
+        }
+
+        //Now deal with the case where the species symbol is a concentration, not an amount.
+        ASTNode scaledConcentrationRate(AST_DIVIDE);
+        ASTNode combinedConcentrationRate(AST_MINUS);
+        if (!speciesIsAmount) {
+            //Need to divide by the compartment size:
+            scaledConcentrationRate.addChild(overallRef->deepCopy());
+            ASTNode* compsize = new ASTNode(AST_NAME);
+            compsize->setName(compId.c_str());
+            scaledConcentrationRate.addChild(compsize);
+            overallRef = &scaledConcentrationRate;
+
+            //If the compartment changes in time, it becomes even more complicated!
+            if (!compIsConst) {
+                //The equation is: rateOf([S1]) = rateOf(S1)/C - [S1]/C * rateOf(C)
+                combinedConcentrationRate.addChild(overallRef->deepCopy());
+                ASTNode* subdiv = new ASTNode(AST_DIVIDE);
+                ASTNode* mult = new ASTNode(AST_TIMES);
+                ASTNode* species = new ASTNode(AST_NAME);
+                species->setName(name.c_str());
+                ASTNode* rateOfComp = new ASTNode(AST_FUNCTION_RATE_OF);
+                ASTNode* comp = new ASTNode(AST_NAME);
+                comp->setName(compId.c_str());
+                rateOfComp->addChild(comp);
+                mult->addChild(rateOfComp);
+                mult->addChild(species);
+                subdiv->addChild(mult);
+                subdiv->addChild(comp->deepCopy());
+                combinedConcentrationRate.addChild(subdiv);
+                overallRef = &combinedConcentrationRate;
+            }
+
+        }
+        //string formula = SBML_formulaToL3String(overallRef);
+        rate = ASTNodeCodeGen(builder, resolver, ctx, modelData).codeGenDouble(overallRef);
+    }
+    else if (dataSymbols.hasRateRule(name))
+    {
+        //NOTE:  if we stored rates persistently, we could use the following instead:
+        //rate = mdbuilder.createRateRuleRateLoad(name, name + "_rate");
+        const LLVMModelSymbols& modelSymbols = ctx.getModelSymbols();
+        SymbolForest::ConstIterator i = modelSymbols.getRateRules().find(
+            name);
+        if (i != modelSymbols.getRateRules().end())
+        {
+            rate = ASTNodeCodeGen(builder, resolver, ctx, modelData).codeGenDouble(i->second);
+        }
+    }
+    else if (dataSymbols.hasAssignmentRule(name))
+    {
+        throw_llvm_exception("Unable to define the rateOf for symbol '" + name + "' as it is changed by an assignment rule.");
+    }
+    else
+    {
+        rate = ConstantFP::get(builder.getContext(), APFloat(0.0));
+    }
```

Comment:
What if we initialize 'rate' with zero value (ConstantFP::get(builder.getContext(), APFloat(0.0))) and avoid having this else statement?

Reply:
If we initialize it and then don't use it, I'm not entirely sure the object won't persist anyway.  If that wasn't true, this would be a good idea.